multi: introduce v2transport and implement BIP324

Crypt-iQ opened the pull request:
Depends on https://github.com/btcsuite/btcd/pull/2219

This PR implements the [BIP324 encrypted version 2 handshake](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki). The service flag is added to the list of supported services so that btcd will optimistically try to make v2 connections.

saubyk commented:
cc: @kcalvinalvin for review

kcalvinalvin commented:
Haven’t yet taken a look at other parts in details but seems like the PR is missing `NODE_P2P_V2` signaling support. Not sure if this part of the code was out of scope for the PR but

Should add

```

SFNodeP2PV2 = 1 << 11

```

here: https://github.com/btcsuite/btcd/blob/24eb815168f49dea84767817717a11bd7928eb23/wire/protocol.go#L72-L105

And other necessary code at:

https://github.com/btcsuite/btcd/blob/24eb815168f49dea84767817717a11bd7928eb23/server.go#L2723-L2743

kcalvinalvin reviewed (commented):
Initial round of review. Went through BIP324 and compared the flows and the pseudocode provided in the BIP with the code in `transport.go` and `chacha.go`

litecomb commented:
WriteV2MessageN should allow specify the garbage []byte, if nil default garbage is used
ReadV2MessageN should allow reading the garbage []byte

Rationale: this will permit custom nonstandard packets communication in the protocol, while still being compatible with bitcoin.

litecomb commented:
It would be great if v2transport wasn't promoted to go module to ease reviewing

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+		if err != nil {
+			return err
+		}
+
```
I believe this is implemented on the receiving side

Crypt-iQ replied:
I think this is implemented properly because we'll continue to parse the 16 bytes and eventually call `CompleteHandshake` which will compare the already-parsed 16 bytes and error out with `errWrongNetV1Peer`. Ideally it would have a test though.

Roasbeef replied:
AFAICT, it's a bit hidden in the pseudo code, it makes V1 prrefix which includes the network magic, see `complete_handshake`: 
> NETWORK_MAGIC = b'\xf9\xbe\xb4\xd9' # Mainnet network magic; differs on other networks.
> V1_PREFIX = NETWORK_MAGIC + b'version\x00\x00\x00\x00\x00'

>     if not initiating and ellswift_theirs[4:16] == V1_PREFIX[4:16]:
>         # Looks like a v1 peer from the wrong network.
>         disconnect(peer)

Crypt-iQ commented:
Pushed up the latest set of changes, will work on addressing the comments. Inbound v2 connections now work. The one thing that is not implemented that bitcoind implements and the BIP text [recommends](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#signaling-specification) we implement is optimistically reconnecting via v1 if an outbound v2 connection failed.

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+
+	var receivedPrefix []byte
+	if initiating {
+		receivedPrefix = make([]byte, 0)
```
I've pre-allocated now, but I think this case should actually be removed?

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+		return errInsufficientBytes
+	}
+
+	if !initiating && bytes.Equal(ellswiftTheirs[4:16], v1Prefix[4:16]) {
```
I think this is fine as-is? Maybe I'm misunderstanding, but I think we'll disconnect the node if they have the wrong network magic? Ideally this would also have a test.

Roasbeef replied:
This matches the logic in `complete_handshake` for the psuedo code in the BIP.

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+	aad := p.sentGarbage
+	for i := 0; i < len(decoyContentLens); i++ {
+		decoyContent := make([]byte, decoyContentLens[i])
+		encPacket, _, err := p.V2EncPacket(decoyContent, aad, false)
```
Yup, nice catch. This code is currently unused but should use a test.

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+	}
+
+	// Send version packet.
+	_, _, err = p.V2EncPacket(transportVersion, aad, false)
```
This needs to be `aad` in case we haven't sent decoy packets which we don't currently

Crypt-iQ commented on peer/peer.go:
```diff
@@ -1067,8 +1074,32 @@ func (p *Peer) handlePongMsg(msg *wire.MsgPong) {
+			plaintext, p.ProtocolVersion(), encoding,
+		)
+		n = len(plaintext)
+		atomic.AddUint64(&p.bytesReceived, uint64(n))
```
good catch!

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+	if err != nil {
+		return err
+	}
+	for i := 0; i < 4096; i++ {
```
Added a comment, I think this should actually be `4095`

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+		ignoreNum = 1
+	}
+
+	ignoreNum <<= ignoreBitPos
```
Done, let me know if this is what you had in mind

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+		return err
+	}
+
+	p.sentGarbage = make([]byte, garbageLen)
```
Yup, now asserting that it's <= 4095 bytes

Crypt-iQ commented on v2transport/chacha.go:
```diff
@@ -0,0 +1,173 @@
+func NewFSChaCha20(initialKey []byte) (*FSChaCha20, error) {
+	var initialNonce [12]byte
+	numRekeys := 0
+	binary.LittleEndian.PutUint64(initialNonce[4:12], uint64(numRekeys))
```
I think this was used in an earlier iteration of the code, removed.

Crypt-iQ commented on v2transport/transport.go:
```diff
@@ -0,0 +1,570 @@
+		return nil, 0, err
+	}
+
+	encPacket := make([]byte, 0)
```
Done

Crypt-iQ commented on v2transport/chacha.go:
```diff
@@ -0,0 +1,173 @@
+	// XOR the text with the keystream to get either the cipher or plaintext.
+	textLen := len(text)
+	dst := make([]byte, textLen)
+	f.cipher.XORKeyStream(dst, text)
```
Yeah, we don't actually need `block_counter` here. I implemented it with `block_counter` and found that it actually served no use since the `XORKeyStream` handles the keystream bytes for us.

Roasbeef commented on v2transport/chacha.go:
```diff
@@ -0,0 +1,172 @@
+	return result, nil
+}
+
+// FSChaCha20 ...
```
Missing `godoc` comments.

Roasbeef replied:
Should mention the purpose of these wrappers. Eg: this one is used to encrypt the lengths of all packets.

Crypt-iQ replied:
Done

Roasbeef commented on v2transport/chacha.go:
```diff
@@ -0,0 +1,172 @@
+
+		var nonce [12]byte
+		numRekeys := f.chunkCtr / rekeyInterval
+		binary.LittleEndian.PutUint64(nonce[4:12], numRekeys)
```
Hmm, I don't see a section of the BIP that matches this.

Crypt-iQ replied:
This is an implementation detail specific to the golang `chacha20` library. We don't get the keystream bytes directly and instead have to use `XORKeyStream` and xor it with empty `[N]byte` to output the keystream bytes. The only time we need to change the nonce here is during each re-keying interval as o/w it stays the same.

Roasbeef commented on v2transport/transport.go:
```diff
@@ -0,0 +1,625 @@
+	sendP *FSChaCha20Poly1305
+
+	// sendGarbageTerm is the garbage terminator that we send.
+	sendGarbageTerm []byte
```
IIUC this is always 16 bytes? So can make this a fixed size array.

Crypt-iQ replied:
Yup

Roasbeef commented on v2transport/transport.go:
```diff
@@ -0,0 +1,625 @@
+}
+
+// createV2Ciphers constructs the packet-length and packet encryption ciphers.
+func (p *Peer) createV2Ciphers(ecdh_secret []byte, initiating bool,
```
ecdh_secret -> ecdhSecret

Roasbeef commented on v2transport/transport.go:
```diff
@@ -0,0 +1,625 @@
+	}
+
+	// TODO:
+	//     To achieve forward secrecy we must wipe the key material used to initialize the ciphers:
```
I think in go we can mainly w/ best effort basis:
  * use something like `mlock` to prevent the secret data from swapping to disk 
  * make functions to zero out bytes either by doing each byte, or setting the pointer to an empty value
      * optionally force a GC at this point 

Can be wroth checking if the crypto stuff in the stdlib does anything fancy here.

Roasbeef replied:
Looks like there's this issue: https://github.com/golang/go/issues/21865

Roasbeef commented on v2transport/transport.go:
```diff
@@ -0,0 +1,625 @@
+// well as garbage data to our peer.
+func (p *Peer) InitiateV2Handshake(garbageLen int) error {
+	var err error
+	p.privkeyOurs, p.ellswiftOurs, err = btcec.EllswiftCreate()
```
Thought this was in a sub-package? If not, we should move it to one.

Crypt-iQ replied:
Will move

Roasbeef commented on v2transport/transport.go:
```diff
@@ -0,0 +1,625 @@
+}
+
+// V2EncPacket takes the contents and aad and returns a ciphertext.
+func (p *Peer) V2EncPacket(contents []byte, aad []byte, ignore bool) ([]byte,
```
Would be good to create some basic benchmarks for:
  * handshake creation
  * handshake complete 
  * v2 encrypt 
  * v2 decrypt 

Just so we can get an ideal re the current cpu/memory utilization.

Roasbeef commented on peer/peer.go:
```diff
@@ -1065,10 +1072,42 @@ func (p *Peer) handlePongMsg(msg *wire.MsgPong) {
+			plaintext, p.ProtocolVersion(), encoding,
+		)
+		n = len(plaintext)
+	} else if partial {
```
?

Crypt-iQ replied:
This is used for when we expect a v2 inbound connection but instead receive a v1 version message. In this case, we've already read some of the v1 version header and need to process the rest of it. Without this, the v2->v1 inbound downgrade doesn't work properly.

Roasbeef replied:
Hmm, ok will examine this relevant area of the BIP further. 

Can we add a comment here explaining the rationale here a bit?

Roasbeef commented on peer/peer.go:
```diff
@@ -2215,7 +2270,39 @@ func (p *Peer) waitToFinishNegotiation(pver uint32) error {
+	downgradedConn := false
+
+	if p.cfg.UsingV2Conn {
+		// TODO: Change garbageLen to be random.
```
So just a random 12 byte int?

Crypt-iQ replied:
I'm not sure why I chose it to be `1` here, it can instead be `0` I think. The garbage here is limited to `4095` bytes.

Crypt-iQ commented on btcec/ellswift.go:
```diff
@@ -0,0 +1,401 @@
+// public key (u, t) and return the point on the curve it maps to. This
+// function returns an error if there is no valid x-coordinate.
+//
+// TODO: Rewrite these to avoid new(FieldVal).Add(...) usage?
```
I forget what I meant with this comment, I think I wanted to see if it was possible to initialize the `FieldVal` with the value instead of calling `new` followed by `Add`?

Crypt-iQ commented on btcec/ellswift.go:
```diff
@@ -0,0 +1,401 @@
+//
+// TODO: Rewrite these to avoid new(FieldVal).Add(...) usage?
+// NOTE: u, t MUST be normalized. The result x is normalized.
+func XSwiftEC(u, t *FieldVal) (*FieldVal, error) {
```
I double-checked and it looks weird, but this is the proper code for `XSwiftEC` from the BIP. Just curious, what made you think that they were combined here?

Crypt-iQ commented on btcec/ellswift.go:
```diff
@@ -0,0 +1,401 @@
+
+	if caseNum&2 == 0 {
+		// If lift_x(-x-u) succeeds, return None
+		if _, found := liftX(new(FieldVal).Add(x).Add(u).Negate(2)); found {
```
The `2` is the magnitude of the value we are negating -- I think this is correct based on my reading of the `Negate` godoc.

Crypt-iQ commented on btcec/ellswift.go:
```diff
@@ -0,0 +1,401 @@
+// isXOnCurve returns true if there is a corresponding y-value for the passed
+// x-coordinate.
+func isXOnCurve(x *secp256k1.FieldVal) bool {
+	y := new(FieldVal).Add(x).Square().Mul(x).AddInt(7)
```
I thought so initially, but I don't think so in this case? From my reading of the godocs on magnitude in the various `FieldVal` functions, `isOnCurve` needs to normalize because `Equals` is being called which requires both inputs to be normalized. With this function `isXOnCurve`, I don't think we need to normalize since none of the used `FieldVal` functions require us to normalize unless the input magnitude is too large.

Crypt-iQ commented on btcec/ellswift.go:
```diff
@@ -0,0 +1,401 @@
+	}
+
+	// Negate y if it's odd.
+	if !y.Negate(1).Normalize().IsOdd() {
```
Ah yeah, this does rely on mutation, @kcalvinalvin pointed out that we can just use `DecompressY` instead. I'll change the code to reflect that

Crypt-iQ replied:
Hmm, on second thought. IIUC `DecompressY` also mutates `resultY` and the reason we want `liftX` is that it returns `PublicKey`. Should I still change this code to use `DecompressY`?

Roasbeef commented:
We should likely have a config to disabled the new behavior as well.

Roasbeef commented on server.go:
```diff
@@ -2209,6 +2209,19 @@ func (s *server) outboundPeerConnected(c *connmgr.ConnReq, conn net.Conn) {
+	// the underlying Peer, trigger a reconnect using the OG v1 connection
+	// scheme.
+	if !sp.Inbound() && sp.Peer.ShouldDowngradeToV1() {
+		// TODO: Determine _how_ to trigger a reconnect that does not use v2
```
I think we can annotate `serverPeer` or wrap it in a struct that adds extra context. 

Then when we go to handle the done peer: https://github.com/btcsuite/btcd/blob/85de89db5b5dd6399e7ff8571ebbf6baece7606b/server.go#L1832-L1842

We annotate a new map (one route) that tracks downgrade attempts that need to be achieved. When we connect to the peer, and go to make a new peer struct, we check that map, and based on that value we set the v2 option in the config.

Crypt-iQ commented:
Pushed up a patch that implements the reconnection logic. It is not currently working on mainnet, I believe it is because I am catching `EOF` instead of some of the other read errors like "use of closed network connection".

Crypt-iQ commented:
Latest set of changes allow the outbound v2->v1 connection downgrade to work properly. Have run this on mainnet and btcd is able to reconnect when an outbound v2 fails.

Roasbeef commented on peer/peer.go:
```diff
@@ -293,6 +293,7 @@ type Config struct {
 
 	// UsingV2Conn is defined if and only if we accept and attempt to make
 	// v2 connections.
+	// TODO: Modify this so that we can downgrade certain peers.
```
WIP TODO?

Roasbeef commented on peer/peer.go:
```diff
@@ -2345,13 +2347,19 @@ func (p *Peer) negotiateOutboundProtocol() error {
 		// bitcoind uses.
 		// TODO: random value for garbage len?
-		if err := p.V2Transport.InitiateV2Handshake(1); err != nil {
+		if err := p.V2Transport.InitiateV2Handshake(0); err != nil {
```
Let's make this a constant somewhere so this section is more readable.

Roasbeef replied:
Also why this change?

Roasbeef replied:
Swapped in to use a proper random garbage length.

Roasbeef commented on server.go:
```diff
@@ -255,6 +255,12 @@ type server struct {
+
+	// pendingReconnects is a map of peer addresses that have failed when
+	// connecting outbound.
+	// TODO: Memory leak if outbound target reached and non-persistent peer.
```
Let's make this into an issue instead? Also what stops of from periodically garbage collecting this? IIUC this map should never have more than outbound target entries

Roasbeef replied:
Can also prob just use `sync.Mutex` to not have to worry as much re critical areas.

Roasbeef commented on server.go:
```diff
@@ -281,6 +287,9 @@ type serverPeer struct {
 	blockProcessed chan struct{}
+
+	reconnectMtx    sync.RWMutex
+	shouldReconnect bool
```
Why do we need this _and_ the field we added in the prior commit? `ShouldDowngradeToV1`

Roasbeef commented on server.go:
```diff
@@ -2187,7 +2218,24 @@ func (s *server) inboundPeerConnected(conn net.Conn) {
-	p, err := peer.NewOutboundPeer(newPeerConfig(sp), c.Addr.String())
+
+	// Just an alias.
+	peerAddr := c.Addr.String()
```
Perhaps we should normalize a bit here and _only_ track the IP? IIUC the port would be included here.

Roasbeef commented on server.go:
```diff
@@ -2214,11 +2262,17 @@ func (s *server) peerDoneHandler(sp *serverPeer) {
+
+		// Flip a bool in serverPeer to mark that we should reconnect to
+		// this peer.
+		sp.SetShouldReconnect()
```
See comment above, re just exposing the internal peer method directly.

Roasbeef commented on connmgr/connmanager.go:
```diff
@@ -463,14 +470,15 @@ func (cm *ConnManager) Connect(c *ConnReq) {
-func (cm *ConnManager) Disconnect(id uint64) {
+// duration. If the reconnect bool is passed, we will attempt to reconnect
+// after the disconnect.
+func (cm *ConnManager) Disconnect(id uint64, reconnect bool) {
```
Is this needed? I think we just actually need to call `Remove`, then we can manually queue the reconnection attempt ourselves.

Roasbeef reviewed (approved):
LGTM 🎄

Pushed up a few commits to help take this across the finish line.

Crypt-iQ reviewed (commented):
Reviewed + tested and it works, ACK 5ab25f3f2bd418fba09325c0f496f56aca03e9a3

I think deployment and cross impl fuzzing will shake out any lingering bugs

Roasbeef commented:
Have been running this without issue for the past week. Confirmed the downgrade logic works for both inbound and outbound peers. Interop with the only other implementation that supports this BIP (`bitcoind`) also is operational. 

Will merge, then create follow up PRs to update the module tags, before a final PR that'll remove the replace directives from the top level module.
